File: src/main/java/ifas/zax/ZaxPublicMorozov.java
package ifas.zax;

import com.zaxsoft.zmachine.ZCPU;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 * Implements reflection hack to avoid changing ZAX code
 */
public class ZaxPublicMorozov {
    private static final Logger LOG = LoggerFactory.getLogger(ZaxPublicMorozov.class);

    public static void zop_save(ZCPU cpu, String dumpFilename, MyZaxUi ui) throws IOException {
        zcpuPersistenceReflectionHelper(cpu, ui, dumpFilename, "zop_save");
    }

    public static void zop_restore(ZCPU cpu, MyZaxUi ui, String dumpFilename) throws IOException {
        zcpuPersistenceReflectionHelper(cpu, ui, dumpFilename, "zop_restore");
    }

    private static void zcpuPersistenceReflectionHelper(ZCPU cpu, MyZaxUi ui, String dumpFilename, String method) {
        synchronized (ui) {
            // makes ZaxUI provide proper filename
            ui.setNextFilename(dumpFilename);
            zcpuReflectionCall(cpu, method);
        }
    }

    public static void zcpuReflectionCall(ZCPU cpu, String method) {
        try {
            Method m = ZCPU.class.getDeclaredMethod(method);
            m.setAccessible(true);

            m.invoke(cpu);
        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
            LOG.error("Could not hack ZAX " + method, e);
        }
    }


    public static void zcpuReflectionFlag(ZCPU cpu, String flag, boolean b) {
        try {
            Field f = ZCPU.class.getDeclaredField(flag);
            f.setAccessible(true);

            f.setBoolean(cpu, b);
        } catch (NoSuchFieldException | IllegalAccessException e) {
            LOG.error("Could not hack ZAX flag " + flag, e);
        }
    }
}

File: src/main/clojure/ifas2/zax_public_morozov.clj
(ns ifas2.zax-public-morozov
  "Uses reflection to access private parts of ZAX. Trading off the ugly hack for unchaned ZAX code."
  (:import
   com.zaxsoft.zmachine.ZCPU

   java.io.IOException
   java.lang.reflect.Field
   java.lang.reflect.InvocationTargetException
   java.lang.reflect.Method)
  (:require [clojure.string :as str]
            [clojure.tools.logging :as log]))

(defn- zcpuReflectionCall [cpu method]
  (try
    (let [m (.getDeclaredMethod ZCPU method (make-array Class 0))]
      (.setAccessible m true)
      (.invoke m cpu (into-array [])))
    (catch NoSuchMethodException e (log/error (str "Could not hack ZAX " method) e))
    (catch InvocationTargetException e (log/error (str "Could not hack ZAX " method) e))
    (catch IllegalAccessException e (log/error (str "Could not hack ZAX " method) e))))

(defn- zcpuReflectionFlag [cpu flag val]
  (try
    (let [f (.getDeclaredField (class ZCPU) flag)]
      (.setAccessible f true)
      (.setBoolean f val))
    (catch NoSuchFieldException e (log/error (str "Could not hack ZAX flag " flag) e))
    (catch IllegalAccessException e (log/error (str "Could not hack ZAX flag " flag) e))))

(defn- zcpuPersistenceReflectionHelper [cpu ui dump-filename method]
  (locking ui

    (.setNextFilename ui dump-filename) ;; makes ZaxUI provide proper filename
    (zcpuReflectionCall cpu method)))

(defn zop-save [cpu ui dump-filename]
  (zcpuPersistenceReflectionHelper cpu ui dump-filename "zop_save"))

(defn zop-restore [cpu ui dump-filename]
  (zcpuPersistenceReflectionHelper cpu ui dump-filename "zop_restore"))

(defn zax-quit [cpu]
  (zcpuReflectionFlag cpu "decode_ret_flag" true)
  (zcpuReflectionCall cpu "zop_quit"))
